Stub out pgrx-internal.md (#1427)
pgcentralfoundation/pgrx: docs/src/contributing/pgrx-internal.md

## pgrx-macros

...whether this is a runtime library or a build tool is a philosophical argument.

# build tools

## pgrx-sql-entity-graph

This crate may seem... somewhat distressing. It is a rather lot of code, some of it without
obvious purpose. One may develop fantasies about deleting it wholesale.  Unfortunately, it serves
an irreplaceable functionality in pgrx, even if its role could be reduced.

A primary task for pgrx, accomplished via cargo-pgrx and support libraries, is to generate SQL.
This allows programmers to avoid separately maintaining the SQL that installs the extension when
`CREATE EXTENSION` runs. This is not merely a convenience: misdeclaring extensions, functions,
types, and operators can in some cases cause illogical behavior from Postgres. The query executor
may assume something based on the SQL definitions, causing it to interact with extensions in ways
that return invalid query results or, depending on the nature of the extension, corrupt a table's
active rows or index. While Postgres employs various techniques to detect such incorrect behaviors
and will abort transactions and roll back if it does, and this may not be "undefined behavior" in

> Rust terms, it is at least as equally undesirable!

Author: Find a way to work in the main citation for this, which is https://www.postgresql.org/docs/current/xoper-optimization.html

Author: Also technically nasal demons are yet-more-powerful in the arbitrariness of the damage they can do, but anything that amounts to risking almost as much damage as `DROP DATABASE production;` is uh... pretty bad!

ways that naturally satisfy this ordering, automated SQL generation technique requires sorting the
SQL declarations to meet these constraints.

Currently, this crate *also* performs the task of reasoning about various lifetime issues in Rust,
and defining some interfaces for translating Rust types into and from SQL. It may be possible to
separate this from the code that maintains aforementioned sorting information, at least somewhat.
Or not.

## cargo-pgrx

Together with `pgrx-sql-entity-graph`, this implements a rather astounding hack:

Various functions are injected into the Rust library, which are then dlopened and called to
extract the required SQL!

See [Forging SQL from Rust](../articles/forging-sql-from-rust.md) for more.

## pgrx-pg-config

This is effectively support library for the build.rs of pgrx-pg-sys and for cargo-pgrx.

## pgrx-version-updater

This tool just helps us update versions for various toml files in the repo, because
the alternative is a horrible pile of bash and regexen.

# test suite

This should be run in CI for every merge.

## pgrx-tests
This currently contains both our test support framework and the actual test suite.

> Fortunately, the way this works is _almost_ magic enough to simply happen if you run `cargo test`.

Author: ```suggestion
Fortunately, the way `#[pg_test]` works is magic enough to simply happen if you run `cargo test`.
```

Unfortunately, due to the way that `#[pg_test]` works, the placement of test code is extremely

> constrained in terms of where it must be in code. Thus we have this additional crate.

Author: ```suggestion
constrained in terms of where it must be in files. This is part of why we have this
additional crate.
```
